### PR TITLE
[Prototype] Add `JavaProxy/JP` to access HDFS

### PR DESCRIPTION
**Pull request by dirtysalt**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [x] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

There are currently a number of operations on BE that involve manipulating Java systems/components, including but not limited to
- Accessing HDFS
- Accessing Hudi MOR tables
- Java UDF
- Access to JDBC
- More Connectors in the future

-----

Currently these implementations are based on JNI。 But JNI implementation has the following problems.
- Stability: The team needs to have people who are proficient in JNI implementation, as well as debugging JVM core issues.
- Development costs: Big data ecosystem is more friendly to Java, and many implementations are available on Java. The cost of porting all these components to JNI is high.
- Development friendly: We will have more connectors in the future. Comparing to C++, Java is more friendly and less error-prone.

-----
So it is proposed to add a separate service `Java Proxy/JP`.
- This Java Proxy is the side car of BE, the purpose is to facilitate BE to access Java system/components.
- Since BE and JP are deployed on same machine, the communication between the two can even bypass serialization or copy. (shared memory or mmap can exchange data between the control flow can use rpc)
- The disadvantage is that it increases the cost of deployment (or BE can fork JP by calling JNI. The good side of this forking way is easy deployment, and the downside is it's hard for monitor and diagnosis by Java tools like jvisualvm)

----
Details & Perf:
- gRPC on Java side
- bRPC on BE side
- http2 as transfer protocol(`h2:grpc`)
- you *have to* increase `h2_client_connection_window_size` (default 1MB, now 64MB), otherwise some RPCs will be blocked

I don't implement bypassing serialization/deser/memcpy between BE & JP now. And I test tpch-100G/Q1 on local machine:
- Direct => 5.14 seconds
- JP => 6.27 seconds
-  JP(mmap) => 5.40 seconds

I guess with some optimization, there is almost no perf penalty.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/fs/fs_hdfs.cpp [hunk 1]
```diff
@@ -3,25 +3,112 @@
 #include "fs/fs_hdfs.h"
 
 #include <fmt/format.h>
-#include <hdfs/hdfs.h>
```
Comment (imay) [inline be/src/fs/fs_hdfs.cpp:6]: Better to create a new filesystem?

**Comment by dirtysalt** [comment]
With mmap optimization, TPCH-100g/Q01 can achieve latency of 5.40sec, which is very close to native one 5.14 sec.

image

File: be/src/fs/fs_hdfs.cpp (hunk 1 above)
Comment (dirtysalt) [reply be/src/fs/fs_hdfs.cpp:6]: Yes, we can abstract this interface to new filesystem.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: java-extensions/java-proxy/src/main/java/com/starrocks/javaproxy/HdfsRpcHandler.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    private Cache<String, CacheValue> cache;
+
+    public HdfsRpcHandler() {
+        Configuration conf = new Configuration();
+        conf.set("fs.hdfs.impl", org.apache.hadoop.hdfs.DistributedFileSystem.class.getName());
+        conf.set("fs.file.impl", org.apache.hadoop.fs.LocalFileSystem.class.getName());
+        configuration = conf;
+
+        cache = CacheBuilder.newBuilder()
+                .maximumSize(128 * 1024)
```
Comment (stdpain) [inline java-extensions/java-proxy/src/main/java/com/starrocks/javaproxy/HdfsRpcHandler.java:49]: suggest make these parameter configurable

Comment (dirtysalt) [reply java-extensions/java-proxy/src/main/java/com/starrocks/javaproxy/HdfsRpcHandler.java:49]: Sure. Right now it's just a prototye.

**Pull request closed without merging**